### ISSUE TITLE
Line breaking around Chinese and Japanese characters are still incorrect

### ISSUE THREAD

**Issue by be5invis** [Microsoft/vscode#330]
When mixing English and Chinese(or Japanese) with whitespace, you can insert line breaks between Chinese(or Japanese) characters, instead of break in whitespace. The current incorrect line breaking can be seen when shrinking the window containing this line 

```天地玄黃宇宙洪荒 English 春夏秋冬風花雪月冠婚葬祭起承轉合喜怒哀樂士農工商生老病死風林火山男女老幼人含道善典謨訓誥```

It should be typesetted as:
![image](https://cloud.githubusercontent.com/assets/240091/11297722/4e1d525c-8fb5-11e5-8abe-8cbc26d48f62.png)

Instead of:
![image](https://cloud.githubusercontent.com/assets/240091/11297738/6e570c70-8fb5-11e5-935b-3900cfe13138.png)

I've checked out source code, by adding **all** Chinese and Japanese characters into a "hidden" setting item `"editor.wordWrapBreakAfterCharacters"`, this issue can be, well, mostly fixed (some Kinsoku Shori is still incorrect, but at least long paragraphs are readable).

**Comment by alexandrudima** [comment]
Reference for a possible improvement in this area:

https://en.wikipedia.org/wiki/Line_breaking_rules_in_East_Asian_languages

**Comment by be5invis** [comment]
@alexandrudima Well to be simple you can separate all Unicode characters into four classes:

* Ideographic: Han, Common Kanas (except small Kanas like ぁ), Hangul.
* Alphabetic: All letters and numbers.
* Leading Character: Opening brackets, like `「` or `(`
* Trailing Character: Commas, Periods or Closing brackets, like `」`, `，` or `)`

The breakability is shown in the table:

First Character →            |Ideographic | Alphabetic | Leading Character | Trailing Character
-------------------------------------|----------------|----------------|--------------------------|--------------------------
Followed by Ideographic     |    yes          |   yes          |        no                    | yes
Followed by Alphabetic       | yes             |  no            |   no                          | yes
Followed by Leading Character | yes      | yes           |       no                       | yes
Followed by Trailing Character | no         | no            | no                              | no

**Comment by be5invis** [comment]
I think I can investigate the `characterHardWrappingLineMapper.ts` to provide a correct CJK-aware wrapper with Kinsoku Shori.

**Comment by be5invis** [comment]
Maybe around line 188:
```
if (charCodeClass === BREAK_AFTER_CLASS) {
	// This is a character that indicates that a break should happen after it
	niceBreakOffset = i + 1;
	niceBreakVisibleColumn = 0;
}
```
We can add another code class `BREAK_IDEOGRAPHIC` and do similar things like this:
```
if (charCodeClass === BREAK_IDEOGRAPHIC && i < len - 1) {
	var nextCode = lineText.charCodeAt(i + 1);
	var nextClass = nextCode < characterClasses.length ? characterClasses[nextCode] : 0;
	if(nextClass !== BREAK_AFTER_CLASS){
		niceBreakOffset = i + 1;
		niceBreakVisibleColumn = 0;
	}
}
```
To initialize `BREAK_IDEOGRAPHIC`, we can simply write this
```
for(i = 0x4E00; i <= 0x9FFF; i++) result[i] = BREAK_IDEOGRAPHIC;
......
```

**Comment by be5invis** [comment]
#4825 should fix this.

**Comment by be5invis** [comment]
I think this issue can be closed, since #4825 has been merged.
